### PR TITLE
fix(cli): large context causes E2BIG error during synthesis on Linux

### LINKED ISSUE #19261: (aws-cdk/core): large context in cdk.context.json causes E2BIG error during synthesis and build
### What is the problem?

We have a service with a large cellular architecture. As we are migrating to CDK from LPT we need to use the Vpc.fromLookup method at least once per account. The size of the generated cdk.context.json exceeds 300K and will only keep getting larger. At this scale the process fails with E2BIG (environment string too big) when [CDK code in this location](https://github.com/aws/aws-cdk/blob/57ad1e087edef653d672c1426b920b12962f0f0f/packages/aws-cdk/lib/api/cxapp/exec.ts#L47) attempts to pass the stringified contents of cdk.context.json to the spawned subprocess via the CDK_CONTEXT_JSON environment.

Environment strings in Linux are generally limited to 128K size and cannot be changed without recompiling the kernel!

MAX_ARG_STRLEN is a constant defined as PAGESIZE*32 in [include/uapi/linux/binfmts.h](https://github.com/torvalds/linux/blob/master/include/uapi/linux/binfmts.h). Its value cannot be changed without recompiling the kernel.

What options do we have to allow using Vpc.fromLookup which is the recommended way to do this and avoid such a roadblock?

### Reproduction Steps

Add sufficient context to your cdk.context.json to exceed the limits, in our case we have a 300K cdk.context.json file. Since this is essentially a hash map of arbitrary values one could fill much of that with garbage random values reach the breaking limit.

Then try to build/synthesize your app which would invoke the CDK_CONTEXT_JSON passing mechanism for exec.

### What did you expect to happen?

Build and synthesize a CDK cloud service.

### What actually happened?

The call will fail with `spawn E2BIG` and process exits with 1.

### CDK CLI Version

1.145.0 (build 9f96380)

### Framework Version

_No response_

### Node.js Version

12

### OS

Amazon Linux 2

### Language

Typescript

### Language Version

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by otaviomacedo**
Linux systems don't support environment variables larger than 128KiB. This change splits the context into two if it's larger than that and stores the overflow into a temporary file. The application then re-constructs the original context from these two sources.

A special case is when this new version of the CLI is used to synthesize an application that depends on an old version of the framework. The application will still consume part of the context, but the CLI warns the user that some of it has been lost.

Since the tree manipulation logic is basically the same as the one used for displaying notices, it was extracted into its own file.

Re-roll #21230
Fixes #19261

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [ ] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [ ] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*
Instead of passing the context in an environment variable, the CLI now writes the context to a temporary file and sets an environment variable only with the location. The app then uses that location to read from the file.

Also tested manually on a Linux machine.

Re-roll #21230
Fixes #19261

### CODE REVIEW

File: packages/aws-cdk/lib/tree.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  readonly children?: { [key: string]: ConstructTreeNode };
+  readonly attributes?: { [key: string]: any };
+
+  /**
+   * Information on the construct class that led to this node, if available
+   */
+  readonly constructInfo?: ConstructInfo;
+}
+
+export function some(node: ConstructTreeNode, predicate: (n: ConstructTreeNode) => boolean): boolean {
```
Comment (comcalvi) [inline packages/aws-cdk/lib/tree.ts:29]: could we get a docstring for this one?

**Review by comcalvi — DISMISSED** [review]
just one small comment. This is pretty much perfect.

**Comment by rix0rrr** [comment]
> Linux systems don't support environment variables larger than 128KiB.

We also have to think about Windows. In Windows, the maximum size of the WHOLE environment together is 32KiB. We can do some calculation based on the current environment to come up with a max for the context value, or do a rough estimation and say we'll cap it at 16KiB or something.

File: packages/aws-cdk/lib/api/cxapp/exec.ts [hunk 2]
```diff
@@ -1,13 +1,19 @@
...
-import { debug } from '../../logging';
+import * as semver from 'semver';
+import { debug, warning } from '../../logging';
 import { Configuration, PROJECT_CONFIG, USER_DEFAULTS } from '../../settings';
+import { loadTree, some } from '../../tree';
+import { splitBySize } from '../../util/objects';
 import { versionNumber } from '../../version';
 import { SdkProvider } from '../aws-auth';
 
+const ENV_VARIABLE_SIZE_LIMIT = 131072; // 128KiB
```
Comment (rix0rrr) [inline packages/aws-cdk/lib/api/cxapp/exec.ts:15]: > Linux systems don't support environment variables larger than 128KiB.

We also have to think about Windows. In Windows, the maximum size of the WHOLE environment together is 32KiB. We can do some calculation based on the current environment to come up with a max for the context value, or do a rough estimation and say we'll cap it at 16KiB or something.

**Review by rix0rrr — CHANGES_REQUESTED** [review]

File: packages/aws-cdk/lib/api/cxapp/exec.ts (hunk 2 above)
Comment (otaviomacedo) [reply packages/aws-cdk/lib/api/cxapp/exec.ts:15]: Good call out!

**Review by rix0rrr — CHANGES_REQUESTED** [review]
I see you've fully switched to the lower of the 2 bounds.

That makes sense, but this will now have the downside that large context setups (anywhere between 32 and 128KiB) and that used to work fine on Linux now stop working in the case of a new CLI and an old framework (which I'm pretty sure will be a large segment of the population).

I think there's not going to be much of a way around having 2 different limits, one for *NIX-based systems and one for Windows, and deciding between them based on the platform we're running on.

**Comment by Naumel** [comment]
@Mergifyio update

**Review by rix0rrr — APPROVED** [review]

**Comment by TheRealAmazonKendra** [comment]
@Mergifyio update

**Pull request merged**